gitpod-io/gitpod
# adding some composition options for modals

## Description
The intention here is to fix the linked issue below so modals are more responsive, and the body of a modal can shrink and scroll when needed. To accomplish that, I needed to add a few css classes to the modal container and body wrapper. In order to allow us to have these sections of a modal consistently rendered, I created a few sub-components so we can compose our Modals a bit as well. `<Modal/>`s can still use the `title` prop and `buttons` prop, which will get wrapped automatically like they were. It might be a good idea to unify how we render modals though at some point, and always compose them like below, but not an urgent change we have to make. These changes should provide a responsive modal w/ scrollable body for any pre-existing modal that was passing in a `title` prop as well.

I've also moved the `classnames` dependency from devDeps to normal deps as I wanted to use it for some conditional class names. I'm happy to adjust that back though if we explicitly don't want to use `classnames` outside of tests though.

```
<Modal ...>
  <ModalHeader ...>
  <ModalContent ...>
  <ModalFooter ...>
</Modal>
```

*We can iterate on this, and add built-in behavior to support pinned error messages on a modal that I saw in at least one place (when adding a new integration). This would allow that style of modal to also render responsively for smaller screen sizes.*

## Related Issue(s)

Fixes #12899, #10747

## How to test
Navigate to [Settings/Integrations](https://bmh-scroll4fb8edc1e5.preview.gitpod-dev.com/integrations) and edit one of the integrations to open the modal. 
* Resize your browser window to be smaller than the height of the modal, and you should see it shrink, while being able to scroll the inner contents of the modal.
<img width="661" alt="image" src="https://user-images.githubusercontent.com/367275/205183812-98e624f1-30fc-4c1f-8bad-117d0c98e7c7.png">
* I've applied the same updates to the Environment Variables modal.


## Release Notes

```release-note
Fixes an issue with modals not displaying properly on smaller screens.
```

## Documentation


## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
